compiler/liftdestructors.nim

```diff
@@ -861,16 +890,16 @@ proc ownedClosureOp(c: var TLiftCtx; t: PType; body, x, y: PNode) =
   #discard addDestructorCall(c, elemType, newNodeI(nkStmtList, c.info), genDeref(xx))
   actions.add callCodegenProc(c.g, "nimDestroyAndDispose", c.info, xx)
   case c.kind
-  of attachedSink, attachedAsgn:
+  of attachedSink, attachedAsgn, attachedDup:
+    if c.kind == attachedDup:
+      body.add genBuiltin(c, mWasMoved, "`=wasMoved`", x)
```

Comment:
It is wrong to treat dup like assignment. And a dup never requires a `wasMoved`, that's the point of dup.

Reply:
Closures aside, How should I implement them for non-ref types?

Reply:
I have removed `wasMoved`. Can I treat them like assignments for non-ref types?

Reply:
No. The way a lifted dup looks like for an object is:

```

proc dup(x: Obj): Obj = Obj(a: dup(x.a), b: dup(x.b)

```

quite different from the other lifted procs. And a challenge for case objects, as usual.

Reply:
Yeah, it is 
```nim
type
  Chuck = object
  Task = object
    id: Chuck
    name: string

proc `=dup`(x: Chuck): Chuck =
  echo "Chuck"
```


```c
N_LIB_PRIVATE N_NIMCALL(tyObject_Task__jHSa1k9c7iuIagzw9b9btXrvQ, eqdup___test54_u92)(tyObject_Task__jHSa1k9c7iuIagzw9b9btXrvQ src_p0)
{
	tyObject_Task__jHSa1k9c7iuIagzw9b9btXrvQ result;
	tyObject_Chuck__kwZ1UmFHwLaBemQzcUHiDA T1_;
	NIM_BOOL *nimErr_;
	{
		nimErr_ = nimErrorFlag();
		nimZeroMem((void *)(&result), sizeof(tyObject_Task__jHSa1k9c7iuIagzw9b9btXrvQ));
		nimZeroMem((void *)(&T1_), sizeof(tyObject_Chuck__kwZ1UmFHwLaBemQzcUHiDA));
		T1_ = eqdup___test54_u5(src_p0.id);
		if (NIM_UNLIKELY(*nimErr_))
			goto BeforeRet_;
		result.id = T1_;
		result.name = eqdup___stdZassertions_u36(src_p0.name);
	}
BeforeRet_:;
	return result;
}
```
The lifted `=dup` hook for `Task` call it's field's `dup` hook.

Reply:
> The way a lifted dup looks like for an object is:

This is how it is lifted now for objects

```nim
proc dup(x: Obj): Obj =
  result.a = `=dup`(x.a)
  result.b = `=dup`(x.b)
```
I assume it's good enough, isn't it?

Reply:
> I assume it's good enough, isn't it?

Only barely. But you can improve it later.

> Is this right for object variants?

Why not `result = Variant(kind: x.kind); case ...`

Reply:
> Why not result = Variant(kind: x.kind); case ...

I don't need to make any changes to generate the former code. Isn't it a pre-existing issue?

Reply:
Yes, you can.